src/trace/tarray.jl

```diff
@@ -63,53 +56,62 @@ function Base.getindex(S::TArray, i::Real)
 end
 
 function Base.setindex!(S::TArray, x, i::Real)
-  t, d = task_local_storage(S.ref)
-  ct = current_task()
+  n, d = task_local_storage(S.ref)
+  cn   = n_copies()
   newd = d
-  if t != ct
+  if cn > n
     # println("[setindex!]: $(S.ref) copying data")
     newd = deepcopy(d)
-    task_local_storage(S.ref, (ct, newd))
+    task_local_storage(S.ref, (cn, newd))
   end
   setindex!(newd, x, i)
 end
 
 function Base.push!(S::TArray, x)
-  t, d = task_local_storage(S.ref)
-  ct = current_task()
+  n, d = task_local_storage(S.ref)
+  cn   = n_copies()
   newd = d
-  if t != ct
+  if cn > n
     newd = deepcopy(d)
-    task_local_storage(S.ref, (ct, newd))
+    task_local_storage(S.ref, (cn, newd))
   end
   push!(newd, x)
 end
 
 function Base.pop!(S::TArray)
-  t, d = task_local_storage(S.ref)
-  ct = current_task()
+  n, d = task_local_storage(S.ref)
+  cn   = n_copies()
   newd = d
-  if t != ct
+  if cn > n
     newd = deepcopy(d)
-    task_local_storage(S.ref, (ct, newd))
+    task_local_storage(S.ref, (cn, newd))
   end
   pop!(d)
 end
 
 function Base.convert(::Type{TArray}, x::Array)
   res = TArray{typeof(x[1]),ndims(x)}();
-  t = current_task()
-  task_local_storage(res.ref, (t,x))
+  n   = n_copies()
+  task_local_storage(res.ref, (n,x))
   res
 end
 
+function Base.convert(::Array, x::Type{TArray})
+  n,d = task_local_storage(S.ref)
+  c = deepcopy(d)
+  return c
+end
+
+
 Base.show(io::IO, S::TArray) = Base.show(io::IO, task_local_storage(S.ref)[2])
 Base.display(io::IO, S::TArray) = Base.display(io::IO, task_local_storage(S.ref)[2])
 Base.size(S::TArray) = Base.size(task_local_storage(S.ref)[2])
 Base.ndims(S::TArray) = Base.ndims(task_local_storage(S.ref)[2])
 
 Base.get(t::Task, S) = S
```

Comment:
This is actually not used. Commented out in https://github.com/yebai/Turing.jl/pull/316.